File: bicycle/car.py
from transportation import Transportation
from typing import Sequence,Set,List,Tuple,Dict
from wheel import Wheel
from transportation import Transportation

class Car(Transportation):
    DEFAULT_FUEL : float = 1.0
    FRONT_LEFT : int = 0
    FRONT_RIGHT : int = 1
    REAR_LEFT : int = 2
    REAR_RIGHT : int = 3
    def __init__(self, diameter : float = Wheel.DEFAULT_DIAMETER, fuel : float = DEFAULT_FUEL):
        super(Car,self).__init__(started = False)
        self._wheels : List[Wheel] = [Wheel(diameter) for i in range(4)]
        self._fuel = fuel

    @property
    def fuel(self) -> float:
        return self._fuel

    @fuel.setter
    def fuel(self,value : float) -> None:
        if value < 0 or value > 1:
            raise ValueError("invalid fuel amount")
        self._fuel = value

    @property
    def frontLeft(self) -> Wheel:
        return self._wheels[self.FRONT_LEFT]

    @property
    def rearLeft(self) -> Wheel:
        return self._wheels[self.REAR_LEFT]
    @property
    def frontRight(self) -> Wheel:
        return self._wheels[self.FRONT_RIGHT]

    @property
    def rearRight(self) -> Wheel:
        return self._wheels[self.REAR_RIGHT]
    @property
    def flat(self) -> bool:
        ans : bool = False
        for wheel in self._wheels:
            ans = ans or wheel.flat
        return ans

# override
    @property
    def going(self):
        return  self.started and not self.flat and self._fuel > 0.0

File: cpp-triangle/src/Car.cpp
#include <stdexcept>
#include "Car.h"
namespace go
{
const double Car::DEFAULT_FUEL = 1.0;
Car::Car(double _diameter)
    : Transportation(false), wheels(4), fuel(DEFAULT_FUEL)
{
  wheels[FRONT_LEFT] = std::shared_ptr<Wheel>(new Wheel(_diameter));
  wheels[FRONT_RIGHT] = std::shared_ptr<Wheel>(new Wheel(_diameter));
  wheels[REAR_LEFT] = std::shared_ptr<Wheel>(new Wheel(_diameter));
  wheels[REAR_RIGHT] = std::shared_ptr<Wheel>(new Wheel(_diameter));
}

double Car::getFuel() const { return fuel; }
void Car::setFuel(const double value)
{
  if (value < 0 || value > 1.0)
  {
    throw std::invalid_argument("invalid fuel amount");
  }
  fuel = value;
}

Wheel &Car::getFrontRight() { return *wheels[FRONT_RIGHT]; }
const Wheel &Car::getFrontRight() const { return *wheels[FRONT_RIGHT]; }

Wheel &Car::getRearRight() { return *wheels[REAR_RIGHT]; }
const Wheel &Car::getRearRight() const { return *wheels[REAR_RIGHT]; }

Wheel &Car::getFrontLeft() { return *wheels[FRONT_LEFT]; }
const Wheel &Car::getFrontLeft() const { return *wheels[FRONT_LEFT]; }

Wheel &Car::getRearLeft() { return *wheels[REAR_LEFT]; }
const Wheel &Car::getRearLeft() const { return *wheels[REAR_LEFT]; }
bool Car::isFlat() const
{
  bool ans = false;
  for (auto wheel : wheels)
  {
    ans = ans || wheel->isFlat();
  }
  return ans;
}

bool Car::isGoing() const { return isStarted() && !isFlat() && fuel > 0.0; }

} // namespace go
